rust/protocol/src/sealed_sender.rs

```diff
@@ -594,96 +598,186 @@ impl UnidentifiedSenderMessage {
+            let derived_values = HKDF::new(3)?.derive_salted_secrets(
+                &shared_secret,
+                &salt,
+                &[],
+                EPHEMERAL_KEYS_KDF_LEN,
+            )?;
 
-            Ok(Self { derived_values })
+            Ok(Self {
+                cipher_key: *array_ref![&derived_values, 32, 32],
+                mac_key: *array_ref![&derived_values, 64, 32],
+            })
         }
+    }
 
-        pub fn cipher_key(&self) -> Result<&[u8]> {
-            Ok(&self.derived_values[32..64])
-        }
+    #[test]
+    fn test_agreement_and_authentication() -> Result<()> {
+        // The sender and recipient each have a long-term identity key pair.
+        let sender_identity = IdentityKeyPair::generate(&mut rand::thread_rng());
+        let recipient_identity = IdentityKeyPair::generate(&mut rand::thread_rng());
+
+        // Generate an ephemeral key pair.
+        let sender_ephemeral = KeyPair::generate(&mut rand::thread_rng());
+        let ephemeral_public = sender_ephemeral.public_key;
+        // Generate ephemeral cipher, chain, and MAC keys.
+        let sender_eph_keys = EphemeralKeys::calculate(
+            &sender_ephemeral,
+            recipient_identity.public_key(),
+            Direction::Sending,
+        )?;
 
-        pub fn mac_key(&self) -> Result<&[u8]> {
-            Ok(&self.derived_values[64..96])
-        }
+        // Encrypt the sender's public key with AES-256 CTR and a MAC.
+        let sender_static_key_ctext = crypto::aes256_ctr_hmacsha256_encrypt(
+            &sender_identity.public_key().serialize(),
+            sender_eph_keys.cipher_key.as_ref(),
+            sender_eph_keys.mac_key.as_ref(),
+        )?;
+
+        // Generate another cipher and MAC key.
+        let sender_static_keys = StaticKeys::calculate(
+            &sender_identity,
+            recipient_identity.public_key(),
+            &sender_eph_keys.chain_key,
+            &sender_static_key_ctext,
+        )?;
+
+        let sender_message_contents = b"this is a binary message";
+        let sender_message_data = crypto::aes256_ctr_hmacsha256_encrypt(
+            sender_message_contents,
+            sender_static_keys.cipher_key.as_ref(),
+            sender_static_keys.mac_key.as_ref(),
+        )?;
+
+        // The message recipient calculates the ephemeral key and the sender's public key.
+        let recipient_eph_keys = EphemeralKeys::calculate(
+            &recipient_identity.into(),
+            &ephemeral_public,
+            Direction::Receiving,
+        )?;
+        assert_eq!(sender_eph_keys, recipient_eph_keys);
+
+        let recipient_message_key_bytes = crypto::aes256_ctr_hmacsha256_decrypt(
+            &sender_static_key_ctext,
+            recipient_eph_keys.cipher_key.as_ref(),
+            recipient_eph_keys.mac_key.as_ref(),
+        )?;
+        let sender_public_key = PublicKey::try_from(recipient_message_key_bytes.as_ref())?;
```

Comment:
Hmmm, I get this error when I try to use a bare `&`:
```rust
error[E0277]: the trait bound `curve::PublicKey: From<Vec<u8>>` is not satisfied
   --> rust/protocol/src/sealed_sender.rs:757:44
    |
757 |         let sender_public_key: PublicKey = PublicKey::try_from(recipient_message_key_bytes)?;
    |                                            ^^^^^^^^^^^^^^^^^^^ the trait `From<Vec<u8>>` is not implemented for `curve::PublicKey`
```

Reply:
Ah, right. You could spell it `&foo[..]` instead of `foo.as_ref()` but it's not the obvious change the others were. (I wish there were a Clippy lint for this, so I wasn't guessing from looking at the code.)

Reply:
I will note that! I have always wanted to make a clippy lint :)))